pkg/types/validation/installconfig.go

```diff
@@ -99,6 +102,123 @@ func ValidateInstallConfig(c *types.InstallConfig, openStackValidValuesFetcher o
+	if n == nil {
+		return
+	}
+	addresses = make(IPByField)
+	for _, network := range n.MachineNetwork {
+		addresses["machineNetwork"] = append(addresses["machineNetwork"], network.CIDR.IP)
+	}
+	for _, network := range n.ServiceNetwork {
+		addresses["serviceNetwork"] = append(addresses["serviceNetwork"], network.IP)
+	}
+	for _, network := range n.ClusterNetwork {
+		addresses["clusterNetwork"] = append(addresses["clusterNetwork"], network.CIDR.IP)
+	}
+	presence = make(IPAddressTypeByField)
+	for k, ips := range addresses {
+		sort.Slice(ips, func(i, j int) bool { return bytes.Compare(ips[i], ips[j]) < 0 })
+		for _, ip := range ips {
+			has := presence[k]
+			if ip.To4() != nil {
+				has.IPv4 = true
+				hasIPv4 = true
+			} else {
+				has.IPv6 = true
+				hasIPv6 = true
+			}
+			presence[k] = has
+		}
+	}
+	return
+}
+
+func ipSliceToStrings(ips []net.IP) []string {
+	var s []string
+	for _, ip := range ips {
+		s = append(s, ip.String())
+	}
+	return s
+}
+
+func ipnetworksToStrings(networks []ipnet.IPNet) []string {
+	var diag []string
+	for _, sn := range networks {
+		diag = append(diag, sn.String())
+	}
+	return diag
+}
+
+// validateNetworkingIPVersion checks parameters for consistency when the user
+// requests single-stack IPv6 or dual-stack modes.
+func validateNetworkingIPVersion(c *types.InstallConfig) field.ErrorList {
+	var allErrs field.ErrorList
+
+	hasIPv4, hasIPv6, presence, addresses := InferIPVersionFromInstallConfig(c.Networking)
+
+	switch {
+	case hasIPv4 && hasIPv6:
+		if c.Networking.NetworkType == "OpenShiftSDN" {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("networking", "networkType"), c.Networking.NetworkType, "dual-stack IPv4/IPv6 is not supported for this networking plugin"))
+		}
+
+		if len(c.Networking.ServiceNetwork) != 2 {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("networking", "serviceNetwork"), strings.Join(ipnetworksToStrings(c.Networking.ServiceNetwork), ", "), "when installing dual-stack IPv4/IPv6 you must provide two service networks, one for each IP address type"))
+		}
+
+		switch {
+		case c.Platform.Azure != nil:
+		case c.Platform.BareMetal != nil:
+		case c.Platform.None != nil:
+		default:
+			allErrs = append(allErrs, field.Invalid(field.NewPath("networking"), "DualStack", "dual-stack IPv4/IPv6 is not supported for this platform, specify only one type of address"))
+		}
+		for k, v := range presence {
+			switch {
+			case k == "machineNetwork" && c.Platform.AWS != nil:
+				// AWS can default an ipv6 subnet
+			case v.IPv4 && !v.IPv6:
+				allErrs = append(allErrs, field.Invalid(field.NewPath("networking", k), strings.Join(ipSliceToStrings(addresses[k]), ", "), "dual-stack IPv4/IPv6 requires an IPv6 address in this list"))
+			case !v.IPv4 && v.IPv6:
+				allErrs = append(allErrs, field.Invalid(field.NewPath("networking", k), strings.Join(ipSliceToStrings(addresses[k]), ", "), "dual-stack IPv4/IPv6 requires an IPv4 address in this list"))
```

Comment:
I hit this validation the first time I tried to use this, using this network config:

```
networking:
  clusterNetwork:
  - cidr: fd01::/48
    hostPrefix: 64
  machineNetwork:
  - cidr: 10.0.0.0/16
  - cidr: ffd0::/48
  networkType: OVNKubernetes
  serviceNetwork:
  - fd02::/112
```

which I think is what we'll need to specify to test single-stack on azure